### ISSUE TITLE
k8s: Pipes/controls don't work most of the time

### ISSUE THREAD

**Issue by 2opremio** [weaveworks/scope#1040]
See video: http://recordit.co/8BTwXcjnRO

It may (or may not) be k8s-specific since it works in my local k8s cluster.

Apart from the unexpected EOF error I also got a `Bad request` in pipes and a few `Probe _ is not connected right now...`

<img width="1772" alt="screen shot 2016-02-25 at 17 33 22" src="https://cloud.githubusercontent.com/assets/2362916/13328380/351dc6c4-dbe6-11e5-91b1-d68890df9699.png">
<img width="1770" alt="screen shot 2016-02-25 at 17 35 28" src="https://cloud.githubusercontent.com/assets/2362916/13328383/396fa440-dbe6-11e5-819a-30b0167005f5.png">

**Comment by 2opremio** [comment]
I repeated the same process while looking at the probe logs and I found this:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x5075bc]

goroutine 5580 [running]:
github.com/weaveworks/scope/probe/docker.(*registry).execContainer(0xc82062a980, 0xc8203b1cc0, 0x40, 0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/docker/controls.go:120 +0x36c
github.com/weaveworks/scope/probe/docker.(*registry).(github.com/weaveworks/scope/probe/docker.execContainer)-fm(0xc8203b1cc0, 0x40, 0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/docker/controls.go:160 +0xca
github.com/weaveworks/scope/probe/docker.captureContainerID.func1(0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/docker/controls.go:149 +0x26d
github.com/weaveworks/scope/probe/controls.HandleControlRequest(0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/weaveworks/scope/probe/controls/controls.go:23 +0x28a
github.com/weaveworks/scope/common/xfer.ControlHandlerFunc.Handle(0x197c9c8, 0xc820de6ca0, 0x10, 0xc8203b1cc0, 0x4c, 0xc821b2b300, 0x15, 0xc820952080, 0x0, 0x0)
        /go/src/github.com/weaveworks/scope/common/xfer/controls.go:48 +0x8b
reflect.Value.call(0x140f5c0, 0x14e2d00, 0x13, 0x16b8628, 0x4, 0xc820647ee8, 0x3, 0x3, 0x0, 0x0, ...)
        /usr/local/go/src/reflect/value.go:432 +0x120a
reflect.Value.Call(0x140f5c0, 0x14e2d00, 0x13, 0xc820647ee8, 0x3, 0x3, 0x0, 0x0, 0x0)
        /usr/local/go/src/reflect/value.go:300 +0xb1
net/rpc.(*service).call(0xc820298100, 0xc8202980c0, 0xc8202bc090, 0xc82000a580, 0xc8202ba140, 0x14e6c20, 0xc821324120, 0xd9, 0x11905a0, 0xc820952080, ...)
        /usr/local/go/src/net/rpc/server.go:383 +0x1c1
created by net/rpc.(*Server).ServeCodec
        /usr/local/go/src/net/rpc/server.go:477 +0x4ac
```

**Comment by 2opremio** [comment]
https://github.com/weaveworks/scope/blob/master/probe/docker/controls.go#L120

Could it be that `exec` is nil?

**Comment by 2opremio** [comment]
Yep, that was the case

```
local = struct { io.Reader; io.Writer }{Reader:(*io.PipeReader)(0xc82002c048), Writer:(*io.PipeWriter)(0xc82002c0c0)}
exec = (*docker.Exec)(nil)
r.client = &docker.Client{SkipServerVersionCheck:true, HTTPClient:(*http.Client)(0xc8202344b0), TLSConfig:(*tls.Config)(nil), Dialer:(*net.Dialer)(0xc820358230), endpoint:"unix:///var/run/docker.sock", endpointURL:(*url.URL)(0xc82060c900), eventMonitor:(*docker.eventMonitoringState)(0xc8205ec3c
0), requestedAPIVersion:docker.APIVersion(nil), serverAPIVersion:docker.APIVersion(nil), expectedAPIVersion:docker.APIVersion(nil), unixHTTPClient:(*http.Client)(0xc8202346c0)}
```

**Comment by 2opremio** [comment]
OK, there are two problems:

1. The crash, which is fixed by #1043
2. The UI is telling the wrong probes to exec into containers. 

Request http://weave-scope-app.default.svc.cluster.local:4040/api/control/4a834abb55fa05eb/076e79e735b4b0a744f6560b1751ea877bbbb0e0153ed54a955e19411da528b5%3B%3Ccontainer%3E/docker_exec_container is targeting probe `4a834abb55fa05eb` but container `076e79e735b4b0a744f6560b1751ea877bbbb0e0153ed54a955e19411da528b5` is running in the machine with probe ID `6c20fcaaa6304b27`: 

```
ubuntu@ip-172-20-0-149:~$ sudo docker logs  $(sudo docker ps | grep scope-probe | grep -v pause | awk '{ print $1 }')  |& grep 'ID'
<probe> INFO: 2016/02/25 22:59:07.160351 probe starting, version 1bfbf67, ID 6c20fcaaa6304b27
ubuntu@ip-172-20-0-149:~$ sudo docker ps | grep 076e79
076e79e735b4        quay.io/weaveworks/app-mapper:0bf7549                 "/app-mapper -db-uri   37 hours ago        Up 37 hours                                  k8s_app-mapper.8b62b9bf_app-mapper-0bf7549-ureto_default_b1e918c5-dade-11e5-91db-0aa65f393a3f_f2fe1667     
```

The container topology (`http://weave-scope-app.default.svc.cluster.local:4040/api/topology/containers/container%3A076e79e735b4b0a744f6560b1751ea877bbbb0e0153ed54a955e19411da528b5`) shows that the App is at fault: 

```json
{
  "node": {
    "rank": "weaveworks/app-mapper",
    "controls": [
      {
        "probeId": "4a834abb55fa05eb",
        "nodeId": "076e79e735b4b0a744f6560b1751ea877bbbb0e0153ed54a955e19411da528b5;<container>",
        "id": "docker_attach_container",
        "human": "Attach",
        "icon": "fa-desktop"
      },
      {
        "probeId": "4a834abb55fa05eb",
        "nodeId": "076e79e735b4b0a744f6560b1751ea877bbbb0e0153ed54a955e19411da528b5;<container>",
        "id": "docker_exec_container",
        "human": "Exec /bin/sh",
        "icon": "fa-terminal"
      },
```

Here's the full topology: https://gist.github.com/2opremio/e91c7f97f0e846076a69

Here's the full report: [report_wrong_container.json.gz](https://github.com/weaveworks/scope/files/147445/report_wrong_container.json.gz)

**Comment by 2opremio** [comment]
I figured out what was happening. There's a ProbeID clash between the kubernetes reporters and the docker reporters. This is because the kubernetes reporter includes information about containers not living in the same host as the probe: https://github.com/weaveworks/scope/blob/7f384032fd0daa04b1b97dbe92e9b20e9d65a496/probe/kubernetes/reporter.go#L72-L77

Distinguish between reporting probes and controlling probes fixes it, I have updated #1043 accordingly.
